History improvements

poelzi opened the pull request:
The history feature has quite some usability issues. The list gets very long very quickly and the most useful entries are at the bottom of the list.

Edit(ronso0) #6917

* Stop wasting the root window with information that are easily discoverable, but instead display the current playlist at the root entry. This way, the menu does not need to be open to jump to the current list.
* Instead of putting the newest entry at the bottom, use descending order and show the last entries first.
* Only show the last 5 playlists and group later entries by year.

![mixxx-history](https://user-images.githubusercontent.com/66107/89414657-1428b880-d72b-11ea-8cb5-7c78e88c5aee.png)

Todo: highlight group if any child is highlighted

poelzi committed:
Make better use of history root view

Most of the time when accessing the history, the current
history is the most important one to categorize tracks, etc.
Currently the user has to open the menu, scroll to the last entry
to access the current playlist. Instead use the root view
as a shortcut to view the current playlist.

poelzi committed:
Order history in descending order (newest on top)

poelzi committed:
Group history by year

Show the last 5 entries directly and group later entries by year.
This boils down the quite long list to a useable amount.

Swiftb0y commented:
It would also be cool if there was a functionality to remove history playlists containing less than N tracks.

daschuer commented on src/library/trackset/baseplaylistfeature.h:
```diff
@@ -102,11 +102,12 @@ class BasePlaylistFeature : public BaseTrackSetFeature {
     void slotTrackSelected(TrackPointer pTrack);
     void slotResetSelectedTrack();
 
+  protected:
```
Please move this to the other protected region in the header.

daschuer commented on src/library/trackset/setlogfeature.cpp:
```diff
@@ -17,6 +17,8 @@
 #include "widget/wlibrarysidebar.h"
 #include "widget/wtracktableview.h"
 
+const int kNumDirectHistoryEntries = 5;
```
This should be in a anonymous namespace, to hide it from the linker. 
It should be also a constexpr, which is really const without exception.

daschuer commented on src/library/trackset/setlogfeature.cpp:
```diff
@@ -136,21 +141,36 @@ void SetlogFeature::onRightClickChild(const QPoint& globalPos, QModelIndex index
 
 QList<BasePlaylistFeature::IdAndLabel> SetlogFeature::createPlaylistLabels() {
-    QList<BasePlaylistFeature::IdAndLabel> playlistLabels;
+    return QList<BasePlaylistFeature::IdAndLabel>();
```
can we dispose the function now?

poelzi replied:
The prototype requires it to be overridden. If I change the prototype, other derived classes may miss implementation when first implemented.

daschuer replied:
Due to your changes the pure virtual implementation + overriding the 
QModelIndex BasePlaylistFeature::constructChildModel()
Is not longer reasonable. Can you check if it's possible to move both to PlaylistFeature?

daschuer commented on src/library/trackset/setlogfeature.cpp:
```diff
@@ -136,21 +141,36 @@ void SetlogFeature::onRightClickChild(const QPoint& globalPos, QModelIndex index
+  * Use a custom model in the histroy for grouping by year
+*/
+QModelIndex SetlogFeature::constructChildModel(int selected_id) {
+    QList<TreeItem*> data_list;
```
Ths can be moved closer to usage and you may call 
```
data_list.reserve(playlistTableModel.rowCount())
```
To avoid memory reallocation

ronso0 commented:
> It would also be cool if there was a functionality to remove history playlists containing less than N tracks.

... as action in history list context menu > QDialog
`max. tracks` spinbox
`Delete` `Cancel`

Or simply being able to select multiple history items in the sidebar, then use Remove from the existing context menu.

daschuer commented on src/library/trackset/setlogfeature.cpp:
```diff
@@ -161,12 +181,44 @@ QList<BasePlaylistFeature::IdAndLabel> SetlogFeature::createPlaylistLabels() {
+        }
+
+        // Create the TreeItem whose parent is the invisible root item
+        TreeItem* item = new TreeItem(name, id);
```
where is this memory freed? you should consider to use a appropriated smart pointer to handle the lifetime of the object.

daschuer commented on src/library/trackset/setlogfeature.cpp:
```diff
@@ -161,12 +181,44 @@ QList<BasePlaylistFeature::IdAndLabel> SetlogFeature::createPlaylistLabels() {
+            if (i != groups.end() && i.key() == year) {
+                i.value()->insertChild(i.value()->childRows(), item);
+            } else {
+                TreeItem* groupItem = new TreeItem(QString("%1").arg(year), -1);
```
here the same, the lifetime of this object is unclear.

daschuer reviewed (changes requested):
Great, this is really a nice improvement. 
I have left some comments.

ronso0 commented:
also somehow related:
https://bugs.launchpad.net/mixxx/+bug/1766163

poelzi commented:
> It would also be cool if there was a functionality to remove history playlists containing less than N tracks.

I had a patch for this as well. Currently the cleanup for empty playlists is in the destructor of a class. Crashes or ctrl+c on a new session always creates empty playlists.

hacksdump commented on src/library/trackset/setlogfeature.cpp:
```diff
@@ -136,21 +141,36 @@ void SetlogFeature::onRightClickChild(const QPoint& globalPos, QModelIndex index
+  * Purpose: When inserting or removing playlists,
+  * we require the sidebar model not to reset.
+  * This method queries the database and does dynamic insertion
+  * Use a custom model in the histroy for grouping by year
```
Function documentation should also include description for the parameter(s).
```suggestion
  * Use a custom model in the histroy for grouping by year
  * @param selected_id <Insert description here>
```

hacksdump replied:
```suggestion
  * Use a custom model in the history for grouping by year
```

hacksdump commented on src/library/trackset/setlogfeature.cpp:
```diff
@@ -136,21 +141,36 @@ void SetlogFeature::onRightClickChild(const QPoint& globalPos, QModelIndex index
+  * This method queries the database and does dynamic insertion
+  * Use a custom model in the histroy for grouping by year
+*/
+QModelIndex SetlogFeature::constructChildModel(int selected_id) {
```
```suggestion
QModelIndex SetlogFeature::constructChildModel(int selectedId) {
```

hacksdump commented on src/library/trackset/setlogfeature.cpp:
```diff
@@ -136,21 +141,36 @@ void SetlogFeature::onRightClickChild(const QPoint& globalPos, QModelIndex index
+*/
+QModelIndex SetlogFeature::constructChildModel(int selected_id) {
+    QList<TreeItem*> data_list;
+    int selected_row = -1;
```
```suggestion
    int selectedRow = -1;
```

poelzi committed:
Implement requested changes.

Fix ownership of TreeItems which are child elements.
Rename variables, add more docs

daschuer commented:
Do you consider this in a mergeable state now? If yes, please remove the draft state. 

And just a hint for the future: 

> Implement requested changes. 

It is not interesting in some years that the commit was a review request. So it is better just to describe the change and split it into separate commits if you do not find a headline that covers all.

ronso0 commented:
I wanted to have this in my 2.3 builds I use live, so I tried to port it to 2.3 (rather hasty copy/paste): https://github.com/ronso0/mixxx/tree/history-improv-2.3

ronso0 commented:
side note / small UX issue I noticed while testing this:

both regular playlists and history playlists can have numbers in parenthesis appended
* for regular playlists this number indicates the number of tracks each list contains
* for history playlists it's append only for the (n>1)-th playlist of a certain date

This is a bit confusing. Maybe we can use `[ ]` for the number of tracks?

foss- commented:
YYYY-MM-DD_n (n2) where n is number of playlist on same date and n2 is the number of tracks on certain playlist. Using a suffix number for number of playlists on same date and parenthesis for number of tracks on tracklist would provide good readability and restore consistency with other playlists used in mixxx.

poelzi committed:
Better handle nested models in library features

Don't assume the list is linear but nest into child object.
User proper match function for finding index of playlist

poelzi committed:
Delete empty histroy playlists on start

Clear database from empty histroy playlists that accumulate from crashes or
closing mixxx with ctrl+c

poelzi committed:
Use / to seperate same date history entries.

Replace directory seperator with - when exporting playlists.

Holzhaus commented on src/library/treeitemmodel.cpp:
```diff
@@ -161,6 +163,12 @@ TreeItem* TreeItemModel::setRootItem(std::unique_ptr<TreeItem> pRootItem) {
 
+/**
+ * Returns the QModelIndex of the Root element.
+ */
```
Please use `///` instead.

poelzi commented:
I tried [] and other unicode characters and found YYYY-MM-DD / n looked the nicest.
Now the tree is properly set bold for the history entry and it's parent.

There is only one thing left that I would like to have:
If you delete a history entry, it would be nice if next item in the tree is selected after deletion. Currently the model is just redrawn and the tree is collapsed. I tried to use the scrollTo function on the Index, but since the index is based on the library feature model and not the sidebar model, the function crashes mixxx.

poelzi committed:
fix crash when index is obsolete due rebuilding of the model

foss- commented:
@poelzi can you attach a screenshot? Wouldn't that introduce an inconsistency with the way other library areas use n? how is n and n2 handled now?

poelzi commented:
![Screenshot_2020-09-22_01-15-15](https://user-images.githubusercontent.com/66107/93830759-612c0400-fc71-11ea-9cd9-4266a5c7f8c8.png)
I can change it to something different, but this looks nice for me.
Export of a playlist now suggests a name in which the dir separator is replaced with a "-".

poelzi commented:
There is only one bug left, but I think it's upstream. scrollTo(QModelIndex) does not work on the sidebar widget.

Holzhaus commented:
There are also multiple unresolved review comments.

ronso0 commented:
> There is only one bug left, but I think it's upstream. scrollTo(QModelIndex) does not work on the sidebar widget.

scrollTo seems to be broken in general. There are multiple outdated Qt bugs for QTreeView but I didn't yet take the time to create a new one for our context with QTableView as selectRow() is a helpful workaround/aternative for selecting and scrolling to an index.

ronso0 commented:
> I can change it to something different, but this looks nice for me.

I vote for @foss- proposal:
**2020-07-26_n (#)**

ronso0 commented:
we could also have
**2020-07-29 2/3 (#)**

poelzi commented:
> we could also have
> **2020-07-29 2/3 (#)**

That is not practical, because it would require to rename all old entries of the same date.

Holzhaus commented:
What about `YYYY-MM-DD - N (M)`? Personally, I don't like the underscore and the slash in incompatible with file names (when exporting).

foss- commented:
@Holzhaus No strong feelings for either proposal.
`YYYY-MM-DD_N (M)` Idea with was to save space.
`YYYY-MM-DD - N (M)` Your proposal might be more human readable.

poelzi commented:
I also really dislike the underscore. I'm ok with the - , just wanted to not use the same as the date. I fixed the issue with exports in general, every directory seperator is replaced by "-" for all playlist exports.

ronso0 commented:
I just don't fnd the ` - ` to be a good indicator for continous numbering.
what about **`YYYY-MM-DD #N (M)`** , would that work for export on all platforms?

Holzhaus commented:
I guess: https://stackoverflow.com/a/31976060

ronso0 commented:
`#` seems to be okay for MacOS, too:
https://en.wikipedia.org/wiki/Filename#Comparison_of_filename_limitations

poelzi committed:
use # seperator for history extra counter

Holzhaus commented on src/library/treeitemmodel.cpp:
```diff
@@ -129,8 +129,10 @@ QModelIndex TreeItemModel::parent(const QModelIndex& index) const {
     TreeItem *childItem = static_cast<TreeItem*>(index.internalPointer());
     TreeItem *parentItem = childItem->parent();
-    if (parentItem == getRootItem()) {
+    if (parentItem == nullptr) {
```
```suggestion
    if (!parentItem) {
```

poelzi committed:
Fix formatting issues from review

uklotzde commented on src/library/trackset/playlistfeature.h:
```diff
@@ -42,12 +40,12 @@ class PlaylistFeature : public BasePlaylistFeature {
     QString fetchPlaylistLabel(int playlistId) override;
-    void decorateChild(TreeItem* pChild, int playlist_id) override;
+    void decorateChild(TreeItem* pChild, int playlistId) override;
+    virtual QList<IdAndLabel> createPlaylistLabels();
```
`override` instead of `virtual`

uklotzde commented on src/library/trackset/playlistfeature.h:
```diff
@@ -42,12 +40,12 @@ class PlaylistFeature : public BasePlaylistFeature {
-    void decorateChild(TreeItem* pChild, int playlist_id) override;
+    void decorateChild(TreeItem* pChild, int playlistId) override;
+    virtual QList<IdAndLabel> createPlaylistLabels();
+    QModelIndex constructChildModel(int selectedId);
```
`override` is missing

poelzi replied:
no, it's not a override anymore

uklotzde commented on src/library/trackset/setlogfeature.cpp:
```diff
@@ -161,12 +179,52 @@ QList<BasePlaylistFeature::IdAndLabel> SetlogFeature::createPlaylistLabels() {
+            if (i != groups.end() && i.key() == yearCreated) {
+                groupItem = i.value();
+            } else {
+                groupItem = new TreeItem(QString("%1").arg(yearCreated), -1);
```
Use QString::number()

uklotzde commented on src/library/trackset/setlogfeature.cpp:
```diff
@@ -161,12 +179,52 @@ QList<BasePlaylistFeature::IdAndLabel> SetlogFeature::createPlaylistLabels() {
+                itemList.append(groupItem);
+            }
+
+            std::unique_ptr<TreeItem> item (new TreeItem(name, id));
```
auto ... = std::make_unique

uklotzde commented on src/library/trackset/setlogfeature.cpp:
```diff
@@ -390,30 +450,25 @@ void SetlogFeature::slotPlaylistTableRenamed(int playlistId, QString newName) {
-    html.append(QStringLiteral("<p>%1</p>").arg(playlistsSummary4));
-    html.append("</td></tr></table>");
-    return html;
+    return QString();
```
Why has this been removed?

poelzi replied:
Because the root of the the history is not uselessly wasted but shows the current history entry. You don't have to open the history at all to go to the current session. The function can not be removed without changing a lot of other features.

uklotzde replied:
A short comment on why this empty implementation is intendend would help. A comment in the PR is not available to future developers who stumble upon this code.

uklotzde commented on src/library/trackset/setlogfeature.h:
```diff
@@ -20,16 +20,17 @@ class SetlogFeature : public BasePlaylistFeature {
 
   protected:
-    QList<BasePlaylistFeature::IdAndLabel> createPlaylistLabels() override;
+    QModelIndex constructChildModel(int selectedId);
```
Missing `override`

poelzi replied:
```
In file included from /home/poelzi/Projects/mixxx/src/library/library.cpp:32:
/home/poelzi/Projects/mixxx/src/library/trackset/setlogfeature.h:33:17: error: ‘QModelIndex SetlogFeature::constructChildModel(int)’ marked ‘override’, but does not override
     QModelIndex constructChildModel(int selectedId) override;
                 ^~~~~~~~~~~~~~~~~~~

```
Does not work. constructChildModel is unique to PlaylistFeature and SetlogFeature, but this function does not exist in BasePlaylistFeature.

uklotzde replied:
What was the reason for removing it from BasePlaylistModel instead of declaring it as pure virtual? Non-virtual, independent methods of the same name in derived classes are totally valid, but could be confusing.

poelzi replied:
Otherwise I would had to implement createPlaylistLabels() which the setlogfeature does not need. Thats why the two distinct implementations when into their respective child classes.

uklotzde commented on src/library/treeitemmodel.cpp:
```diff
@@ -129,8 +129,10 @@ QModelIndex TreeItemModel::parent(const QModelIndex& index) const {
-    if (parentItem == getRootItem()) {
+    if (!parentItem) {
         return QModelIndex();
+    } if (parentItem == getRootItem()) {
```
Wrong and misleading formatting. Should this have been `else if`?

uklotzde commented on src/library/treeitemmodel.h:
```diff
@@ -37,6 +37,7 @@ class TreeItemModel : public QAbstractItemModel {
     TreeItem* getRootItem() const {
         return m_pRootItem.get();
     }
+    QModelIndex getRootIndex();
```
`const` function

uklotzde commented on src/library/treeitemmodel.cpp:
```diff
@@ -161,6 +163,11 @@ TreeItem* TreeItemModel::setRootItem(std::unique_ptr<TreeItem> pRootItem) {
     return getRootItem();
 }
 
+/// Returns the QModelIndex of the Root element.
```
The doc comments need to be moved into the .h file. In the .cpp file they are useless.

uklotzde commented:
Please enable the pre-commit hooks before committing any code.

poelzi commented on src/library/dao/playlistdao.cpp:
```diff
@@ -208,6 +208,27 @@ void PlaylistDAO::deletePlaylist(const int playlistId) {
 
+void PlaylistDAO::deleteEmptyPlaylists(const PlaylistDAO::HiddenType type) {
+    //qDebug() << "PlaylistDAO::deletePlaylist" << QThread::currentThread() << m_database.connectionName();
+    ScopedTransaction transaction(m_database);
```
Transactions help the database to correctly flush and helps with constancy. Without measuring it, I would even suspect it's faster with then without. As long as we don't have proper transaction infrastructure, I would rather let it in.

uklotzde replied:
SQLite does not support nested transactions. The transaction around a single statement is not only redundant here, it may cause an outer transaction to not work as intended.

poelzi commented:
@uklotzde I tried to but it failed for reasons I don't know yet.

poelzi committed:
fix formatting and use const where possible

poelzi committed:
use older lookup method for sibilings

poelzi committed:
fix formatting

poelzi committed:
Remove transaction from playlist removeal

poelzi committed:
fix build warning on ubuntu

/home/travis/build/mixxxdj/mixxx/src/library/trackset/baseplaylistfeature.cpp: In member function ‘void BasePlaylistFeature::markTreeItem(TreeItem*)’:
/home/travis/build/mixxxdj/mixxx/src/library/trackset/baseplaylistfeature.cpp:699:25: error: suggest parentheses around assignment used as truth value [-Werror=parentheses]

             while (item = item->parent()) {

                    ~~~~~^~~~~~~~~~~~~~~~

cc1plus: all warnings being treated as errors

Holzhaus commented on src/library/trackset/baseplaylistfeature.cpp:
```diff
@@ -689,24 +673,36 @@ void BasePlaylistFeature::slotTrackSelected(TrackPointer pTrack) {
+        pTreeItem->setBold(shouldBold);
+        if (shouldBold && pTreeItem->hasParent()) {
+            TreeItem* item = pTreeItem;
+            while ((item = item->parent())) {
```
They are not redundant, because the compiler will generate a warning without extra parentheses when doing an assignment inside a loop condition. The root issue is that assignments inside loop conditions are confusing and should be avoided IMHO.

ronso0 commented:
> > There is only one bug left, but I think it's upstream. scrollTo(QModelIndex) does not work on the sidebar widget.
> 
> scrollTo seems to be broken in general. There are multiple outdated Qt bugs for QTreeView but I didn't yet take the time to create a new one for our context with QTableView as selectRow() is a helpful workaround/aternative for selecting and scrolling to an index.

Anyway, scrollTo() doesn't seem to be necessary IMO because the to-be-deleted History is in view when clicking on it.
The actual bug is that the next history gets selected (tracks show up in the table) but it's not highlighted (visually selected).
Though this works perfectly for regular playlists (which trigger the same functions in baseplaylistfeature and playlist DAO if I'm not mistaken).

Holzhaus commented:
There are merge conflicts now.

poelzi committed:
simplify code

poelzi committed:
Properly open all submenus when using selectIndex on features

ronso0 commented:
> Anyway, scrollTo() doesn't seem to be necessary IMO because the to-be-deleted History is in view when clicking on it.
The actual bug is that the next history gets selected (tracks show up in the table) but it's not highlighted (visually selected).
Though this works perfectly for regular playlists (which trigger the same functions in baseplaylistfeature and playlist DAO if I'm not mistaken).

@poelzi Could you give this another look?
This branch fixes the re-selection of the following _regular_ playlist after deletion (which doesn't work in `main`) but fails for histories.
Maybe because the tree is rebuild due to the re-population of the top section (last five histories)?

poelzi commented:
@ronso0 I know that but. It's a tricky one but your suggestion is something to look into.
I also found multiple places where it assumes a flat hierarchic structure and fixed those already. some code paths seem to have been broken on the way or never worked at all

ywwg commented:
any progress on this?

ronso0 commented:
Besides the new conflicts this works very well.
The only minor issue that's left is the Playlist selection after having deleted one item.

poelzi commented:
No, was busy with other branches

ronso0 commented:
> Besides the new conflicts this works very well.
> The only minor issue that's left is the Playlist selection after having deleted one item.

When the conflicts are resolve we can merge this IMO (given there are no pending objections regrading the code @uklotzde ?) and fix the selection issue later on.

poelzi committed:
Add functions to scroll to the correct child on the sidebar

Use new utility functions to scroll to the next item after the deleted
one.

poelzi committed:
Use new child selection method for join with previous

poelzi committed:
Restore expansion code

poelzi commented:
@ronso0 After some fighting I found the right place in this mess to poke....
Added some easy to use functions to get any child in the sidebar tree into visible range.
delete and join with previous now behaves nicely.

ronso0 commented:
I can confirm it now scrolls to the previous item (centered in siedebar if possible), but without a selection highlight that makes the situation more confusing IMO. Did a solution for this emerge when you worked on the scrolling?

Btw after renaming or locking a history the subtree is collapsed and the History main item is selected while the renamed playlist is shown in the table.

poelzi committed:
Cleanup child selection code

poelzi committed:
Fix memory leak in selection model

poelzi committed:
Fix join with previous criteria logic

poelzi committed:
Fix foorloop clazy warning

poelzi commented:
@Holzhaus I don't understand the clazy warning here. initActions() is not virtual there and I don't get this error message in my branch with clazy 1.7 or 1.8.
@ronso0 I went through all actions and checked both, I think it behaves now nicely in all cases I could think of.

uklotzde commented:
> @Holzhaus I don't understand the clazy warning here. initActions() is not virtual there and I don't get this error message in my branch with clazy 1.7 or 1.8.
> @ronso0 I went through all actions and checked both, I think it behaves now nicely in all cases I could think of.

clazy: Maybe a false positive caused by slotPlaylistTableChanged() that is used in a lamdba expression?

poelzi committed:
Remove lambda due false positive of clazy

poelzi commented:
@uklotzde you won the jackpot ;)
afaik, this one is ready

uklotzde commented on src/library/dao/playlistdao.cpp:
```diff
@@ -227,6 +227,24 @@ void PlaylistDAO::deletePlaylist(const int playlistId) {
     }
 }
 
+void PlaylistDAO::deleteAllEmptyPlaylists(PlaylistDAO::HiddenType type) {
```
How about a more versatile function `deleteAllPlaylistsWithFewerItems(PlaylistDAO::HiddenType type, int minNumberOfItems)`
and then invoke it with `minNumberOfItems = 1` for this use case?

poelzi replied:
done

uklotzde commented on src/library/trackset/baseplaylistfeature.cpp:
```diff
@@ -151,20 +157,34 @@ int BasePlaylistFeature::playlistIdFromIndex(const QModelIndex& index) {
 }
 
+void BasePlaylistFeature::selectPlaylistInSidebar(int playlistId) {
+    if (playlistId == -1) {
```
```
if (!m_pSidebarWidget) {
    return;
}
```

Or even VERIFY_OR_DEBUG_ASSERT?

poelzi replied:
Does not work due startup

uklotzde commented on src/library/trackset/baseplaylistfeature.cpp:
```diff
@@ -151,20 +157,34 @@ int BasePlaylistFeature::playlistIdFromIndex(const QModelIndex& index) {
         m_pPlaylistTableModel->setTableModel(playlistId);
         emit showTrackModel(m_pPlaylistTableModel);
         emit enableCoverArtDisplay(true);
+        if (m_pSidebarWidget) {
```
Same here, check m_pSidebarWidget and otherwise early return when entering this function.

poelzi replied:
Thats not possible here, because this function is called in init when m_pSidebarWidget is not populated yet.

uklotzde commented on src/library/trackset/baseplaylistfeature.cpp:
```diff
@@ -151,20 +157,34 @@ int BasePlaylistFeature::playlistIdFromIndex(const QModelIndex& index) {
-    //qDebug() << "BasePlaylistFeature::activatePlaylist()" << playlistId;
+    // qDebug() << "BasePlaylistFeature::activatePlaylist()" << playlistId;
     QModelIndex index = indexFromPlaylistId(playlistId);
     if (playlistId != -1 && index.isValid()) {
```
One more, early return if playlistId == -1. I will not add more comments about this issue, please check all new or edited functions.

uklotzde commented on src/library/trackset/baseplaylistfeature.cpp:
```diff
@@ -326,6 +346,10 @@ void BasePlaylistFeature::slotCreatePlaylist() {
+            m_lastRightClickedIndex.sibling(m_lastRightClickedIndex.row() + 1,
+                    m_lastRightClickedIndex.column());
+    int nextId = playlistIdFromIndex(nextIndex);
     if (playlistId == -1) {
```
A similar use case. Check the value playlistId immediately and otherwise early return as soon as possible.

poelzi replied:
nextId can be -1, it's not an error case

uklotzde replied:
I didn't mention nextId.

poelzi replied:
fixed

uklotzde commented on src/library/trackset/baseplaylistfeature.cpp:
```diff
@@ -599,6 +628,11 @@ void BasePlaylistFeature::bindLibraryWidget(WLibrary* libraryWidget,
 
+void BasePlaylistFeature::bindSidebarWidget(WLibrarySidebar* pSidebarWidget) {
+    // store the sidebar widget pointer for later use in onRightClickChild
+    m_pSidebarWidget = pSidebarWidget;
```
`DEBUG_ASSERT(!m_pSidebarWidget);` because initializing this member twice would be unexpected.

uklotzde commented on src/library/trackset/baseplaylistfeature.h:
```diff
@@ -97,6 +104,7 @@ class BasePlaylistFeature : public BaseTrackSetFeature {
     QAction* m_pAnalyzePlaylistAction;
 
     PlaylistTableModel* m_pPlaylistTableModel;
+    QSet<int> m_playlistsSelectedTrackIsIn;
```
The naming is not very intuitive. How about `m_playlistIdsOfSelectedTrack`?

uklotzde commented on src/library/trackset/setlogfeature.cpp:
```diff
@@ -18,6 +18,10 @@
 #include "widget/wtracktableview.h"
 
+namespace {
+constexpr int kNumDirectHistoryEntries = 5;
```
"direct"? Please use a more descriptive naming for this constant that reveals the purpose.

uklotzde commented on src/library/trackset/setlogfeature.cpp:
```diff
@@ -394,28 +451,28 @@ void SetlogFeature::slotPlaylistTableRenamed(int playlistId, const QString& newN
+}
+
+void SetlogFeature::activatePlaylist(int playlistId) {
+    //qDebug() << "BasePlaylistFeature::activatePlaylist()" << playlistId;
```
In this function you don't check m_pSidebarWidget while in BasePlaylistFeature you do.

Just noticed this inconsistency, not sure how to resolve it.

poelzi replied:
activateChild checks if m_pSidebarWidget is set and selects the proper index. The checks are only in startup of relevance.

ronso0 commented:
> @Holzhaus I don't understand the clazy warning here. initActions() is not virtual there and I don't get this error message in my branch with clazy 1.7 or 1.8.
> @ronso0 I went through all actions and checked both, I think it behaves now nicely in all cases I could think of.

Yes, works nicely! The correct items are selected after remove, rename or lock.
There's only one inconsistency left, but that is not an real issue:
when I remove `plop` here the History root item is selected. Not an issue because those items are very close
![image](https://user-images.githubusercontent.com/5934199/100449728-b55dc500-30b4-11eb-8aad-bed5457ddf99.png)

poelzi committed:
Make playlist cleanup more versatile.

Allow to delete playlists with less then n tracks.

poelzi committed:
Use constant for indicating invalid playlist id

poelzi committed:
Try to select a sibiling on the same level as the deleted entry

poelzi committed:
Name cleanup. Don't select on root item

poelzi committed:
eager return in error case

poelzi commented:
@ronso0 it now tries to select a child in the same area of the deleted one. It tries the next and if there is none, selects the previous one.
@uklotzde I incorporated your suggestions.

poelzi committed:
simplify slotDeletePlaylist()

poelzi committed:
simplify code

poelzi commented:
> It would also be cool if there was a functionality to remove history playlists containing less than N tracks.

This can now be implemented very easy ontop of this later.

uklotzde reviewed (changes requested):
The batch deletion on initial startup freezes the splash screen for an extended period of time (up to a few minutes) and spams the log:

```
...
debug [Main] Delete playlist 2812
debug [Main] Committing transaction on "MIXXX-1" result: true
debug [Main] PlaylistDAO::getHiddenType returns PLHT_UNKNOWN for playlistId  2812
debug [Main] PlaylistDAO::getHiddenType returns PLHT_UNKNOWN for playlistId  2812
debug [Main] Delete playlist 2813
debug [Main] Committing transaction on "MIXXX-1" result: true
debug [Main] PlaylistDAO::getHiddenType returns PLHT_UNKNOWN for playlistId  2813
debug [Main] PlaylistDAO::getHiddenType returns PLHT_UNKNOWN for playlistId  2813
debug [Main] Delete playlist 2814
debug [Main] Committing transaction on "MIXXX-1" result: true
debug [Main] PlaylistDAO::getHiddenType returns PLHT_UNKNOWN for playlistId  2814
debug [Main] PlaylistDAO::getHiddenType returns PLHT_UNKNOWN for playlistId  2814
...
```

This behavior has to be fixed before merging, even if it only happens during the first startup. Users will likely kill the application before the cleanup has finished.

uklotzde commented:
Suggestion: "Create new history playlist" -> "Finish current history playlist"

This is what the function actually does. Consecutively a new current history playlist is created and started. First I was confused why this option only appeared on the current (top most) playlist.

Bug: The selection changes to the newly created playlist, but the table view still shows the entries from the previously selected playlist.

ronso0 commented:
> @ronso0 it now tries to select a child in the same area of the deleted one. It tries the next and if there is none, selects the previous one.

Yes, it works perfectly now! Thank you very much for your work!
(except the issue @uklotzde noticed)

poelzi commented:
@uklotzde Damn, I think I remember that I implemented the cleanup function first like now, but forgot how slow it was when you have lots of old entries. I will handcraft something fast.

ronso0 commented:
can the History clean up be a progress dialog at startup like library rescan?

uklotzde commented:
> can the History clean up be a progress dialog at startup like library rescan?

Not needed. This should become a single SQL command and finish in no time.

poelzi committed:
Speedup playlist deletion logic

poelzi committed:
Always show start new playlist and focus on change

poelzi commented:
@uklotzde I think start new playlist only on the first entry is perfectly fine, but I changed it to be always visible on the first entry but disabled when the entry makes no sense. For me, this feels more consistent behavior.
Cleanup should now be fast, was not able to get sqlite to delete from 2 tables at once but to fast deletes was possible.

uklotzde commented on src/library/dao/playlistdao.h:
```diff
@@ -59,6 +59,9 @@ class PlaylistDAO : public QObject, public virtual DAO {
     void deletePlaylist(const int playlistId);
+    /// Delete Playlists with fewer entries then "length"
+    /// @return number of deleted playlists, -1 on error
+    int deleteAllPlaylistsWithFewerItems(PlaylistDAO::HiddenType type, int length);
```
It doesn't hurt to let the parameters speak for themselves: "length" -> "minNumberOfTracks"

Each "item" in a playlist is a track, we do not need to introduce a generic term here: -> deleteAllPlaylistsWithFewerTracks

uklotzde commented on src/library/dao/playlistdao.cpp:
```diff
@@ -227,6 +228,51 @@ void PlaylistDAO::deletePlaylist(const int playlistId) {
+    }
+    deletePlaylists.execPrepared();
+
+    auto deleteTracks = FwdSqlQuery(m_database,
```
Delete the tracks before their playlists. Otherwise referential integrity is violated temporarily.

The corresponding SQLite option is currently disabled, but might be enabled in the future and would let this ordering of statements fail.

uklotzde commented on src/library/dao/playlistdao.cpp:
```diff
@@ -227,6 +228,51 @@ void PlaylistDAO::deletePlaylist(const int playlistId) {
+    }
+    QString idString = idStringList.join(",");
+
+    qDebug() << "Delete Playlists: " << idString;
```
qInfo() << "Deleting" << idStringList.size() << "playlists of type" << type << "that contain fewer than" << minNumerOfTracks << "tracks";

The ids should not matter if the code has been tested thoroughly.

uklotzde commented on src/library/dao/playlistdao.cpp:
```diff
@@ -227,6 +228,51 @@ void PlaylistDAO::deletePlaylist(const int playlistId) {
+int PlaylistDAO::deleteAllPlaylistsWithFewerItems(PlaylistDAO::HiddenType type, int length) {
+    QSqlQuery query(m_database);
+
+    ScopedTransaction transaction(m_database);
```
Move the start of the transaction to the top.

As already mentioned this should be extracted from the DAOs. All DAOs should safely assume that their functions are executed in a proper transaction scope. They should not (need) to make any assumptions about the outer context.

uklotzde replied:
Just noticed that you added this transaction recently:
- Move the transaction out of PlaylistDAO and to SetlogFeature 
- Add a comment to the function docs that states that it has to be executed within a transaction.

Unfortunately, we cannot determine if a transaction has already been started. This would require to introduce a wrapper for QSqlDatabase.

uklotzde commented on src/library/dao/playlistdao.cpp:
```diff
@@ -227,6 +228,51 @@ void PlaylistDAO::deletePlaylist(const int playlistId) {
+
+    qDebug() << "Delete Playlists: " << idString;
+
+    auto deletePlaylists = FwdSqlQuery(m_database,
```
If you delete their tracks first then you could simply delete all empty playlists of the given type as the 2nd step, independent of the parameter `minNumberOfTracks`.

uklotzde commented on src/library/dao/playlistdao.cpp:
```diff
@@ -227,6 +228,51 @@ void PlaylistDAO::deletePlaylist(const int playlistId) {
 }
 
+int PlaylistDAO::deleteAllPlaylistsWithFewerItems(PlaylistDAO::HiddenType type, int length) {
+    QSqlQuery query(m_database);
```
```
VERIFY_OR_DEBUG_ASSERT(minNumberOfTracks > 0) {
    return 0; // nothing to do
}
```

poelzi committed:
Move transaction outside of DAO

poelzi committed:
Print info instead of debug when transaction failed

poelzi committed:
Don't double log on errors

poelzi commented:
I'm not interested doing any more polishing here.

daschuer commented on src/library/trackset/baseplaylistfeature.cpp:
```diff
@@ -16,11 +16,17 @@
 #include "widget/wlibrarytextbrowser.h"
 
+namespace {
+const char* kUnsafeFilenameReplacement = "-";
```
```suggestion
constexpr QChar kUnsafeFilenameReplacement = '-';
```

daschuer commented on src/library/trackset/baseplaylistfeature.cpp:
```diff
@@ -139,44 +145,70 @@ void BasePlaylistFeature::initActions() {
+    emit enableCoverArtDisplay(true);
+    // Update selection
+    emit featureSelect(this, m_lastRightClickedIndex);
+    activateChild(m_lastRightClickedIndex);
```
```suggestion
    if (!m_pSidebarWidget) {
        m_pSidebarWidget->selectChildIndex(index);
    }
```

is this sufficient here?

daschuer commented on src/library/trackset/baseplaylistfeature.cpp:
```diff
@@ -323,10 +355,33 @@ void BasePlaylistFeature::slotCreatePlaylist() {
 
+/// Returns a playlist that is a sibling inside the same parent
+/// as the start index
+int BasePlaylistFeature::selectSiblingPlaylistId(QModelIndex& start) {
```
```suggestion
int BasePlaylistFeature::getSiblingPlaylistIdOf(QModelIndex& start) {
```
we do not select here. The Qt sibling() function verifies if the test position is a sibling. Here you look at all columns below start.

poelzi replied:
It takes the next sibling if it exists, otherwise it checks if the one in front exists and returns an invalid id otherwise. I can understand that select could be misinterpreted in gui context.

daschuer commented on src/library/trackset/baseplaylistfeature.cpp:
```diff
@@ -336,13 +391,13 @@ void BasePlaylistFeature::slotDeletePlaylist() {
-        VERIFY_OR_DEBUG_ASSERT(playlistId >= 0) {
-            return;
-        }
+    int nextId = selectSiblingPlaylistId(m_lastRightClickedIndex);
```
```suggestion
    int siblingId = selectSiblingPlaylistId(m_lastRightClickedIndex);
```

daschuer commented on src/library/trackset/baseplaylistfeature.cpp:
```diff
@@ -336,13 +391,13 @@ void BasePlaylistFeature::slotDeletePlaylist() {
+    int nextId = selectSiblingPlaylistId(m_lastRightClickedIndex);
+
+    m_playlistDao.deletePlaylist(playlistId);
+    activate();
```
It feels wrong to activate a unrelated sibling playlist when currently something else like AutoDj is selected.

daschuer commented:
The "Join with previous" feature is now broken. Out of habit I expected that it deletes the current playlist and joins it with the one above. Unfortunately it now joins with the unrelated day below. 
Can we rename that feature  "Join with previous (below)" or explicit name the playlist joined?

daschuer reviewed (changes requested):
Thank you. Some comments.

poelzi commented:
> The "Join with previous" feature is now broken. Out of habit I expected that it deletes the current playlist and joins it with the one above. Unfortunately it now joins with the unrelated day below.
> Can we rename that feature "Join with previous (below)" or explicit name the playlist joined?

This is just habit and will change once used to it. It behaves correctly here and joins the entry with the previous entry which is below the current one, because it is further in time.

daschuer commented:
> This is just habit and will change once used to it. It behaves correctly here and joins the entry with the previous entry which is below the current one, because it is further in time.

The issue is that it can't be undone. As I hit the trap, many others will do.
So let's at least at least change the name to "Join with previous (below)"

ronso0 commented:
The timeline is now inverted, this feels very intuitive to me. I assume most users have one history per day so the date numbers make that clear instantly IMO. 
So "Join with previous" gives a step learning curve is quite steep ;) I would not add 'below' here.

poelzi committed:
use constexpr when possible. Clearify join wording

poelzi committed:
Don't activate when not expected.

Only activate playlists when already in focus.
Add option to scroll but not select sidebar items.

poelzi commented:
@daschuer Fixed your points

daschuer commented on src/library/trackset/baseplaylistfeature.cpp:
```diff
@@ -16,11 +16,17 @@
 #include "widget/wlibrarytextbrowser.h"
 
+namespace {
+auto kUnsafeFilenameReplacement = QLatin1String("-");
```
```suggestion
constexpr QChar kUnsafeFilenameReplacement = '-';
```
What was the issue with this proposal? 
constexpr ensures that this is a constant in the text segment. 
QChar is already the 16 bit representation of the char, this avoids conversion at runtime.
replace() needs the 16 bit representation. 

It is different story when comparing strings. In this case Qt performs a 8 bit char compare which is faster than a 16 bit compare and a Latin1String() is the fates alternative then.

poelzi replied:
I see, fixed

poelzi committed:
use QChar instead of StringLiteral

uklotzde reviewed (approved):
Works fine so far and is a huge usability improvement.

A nice extension for the future would be relative date folders:

- today
- yesterday
- last week
- last month
- last year
- ...remaining playlists grouped by calendar year...

daschuer reviewed (approved):
LGTM, Thank you.

daschuer commented:
@Holzhaus: merge?

Holzhaus commented:
> @Holzhaus: merge?

Feel free to merge, I didn't follow this PR.

uklotzde commented:
All review comments have been addressed. Thank you! LGTM

ronso0 commented:
:tada:

Be-ing commented:
The meaning of the number after the playlist date is unclear to me. Why does sometimes it have a `#` whereas other times it is in parentheses?

ronso0 commented:
# is the prefix for the Nth playlist on that day.
number in parenthesis is the track count.

Be-ing commented:
I think I got confused because there is no number in parentheses if the playlist only has 1 track.

Holzhaus commented:
> I think I got confused because there is no number in parentheses if the playlist only has 1 track.

I agree, we should always show the track count.

poelzi commented:
I will open a followup
